### PR TITLE
CHG: [droid] use native interface for mediacodec

### PR DESCRIPTION
**Pull request by koying**
## Motivation and Context
Gain of performance and resource by not having to use JNI

## Types of change

- [X] Improvment / Refactor 
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Comment by peak3d** [comment]
@koying is the performance gain / cpu reduction really noticable?

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp [hunk 1]
```diff
@@ -553,14 +568,11 @@ bool CDVDVideoCodecAndroidMediaCodec::Open(CDVDStreamInfo &hints, CDVDCodecOptio
     {
       if (types[j] == m_mime)
       {
-        m_codec = std::shared_ptr<CJNIMediaCodec>(new CJNIMediaCodec(CJNIMediaCodec::createByCodecName(m_codecname)));
-
-        // clear any jni exceptions, jni gets upset if we do not.
-        if (xbmc_jnienv()->ExceptionCheck())
+        m_codec = AMediaCodec_createCodecByName(m_codecname.c_str());
```
Comment (peak3d) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp]: Really sure that we don't need a shared_ptr here?

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp [hunk 2]
```diff
@@ -1165,30 +1101,27 @@ int CDVDVideoCodecAndroidMediaCodec::GetOutputPicture(void)
...
-    CJNIMediaFormat mediaformat = m_codec->getOutputFormat();
-    if (xbmc_jnienv()->ExceptionCheck())
-    {
-      xbmc_jnienv()->ExceptionDescribe();
-      xbmc_jnienv()->ExceptionClear();
-    }
-    ConfigureOutputFormat(&mediaformat);
+    AMediaFormat* mediaformat = AMediaCodec_getOutputFormat(m_codec);
+    if (!mediaformat)
+      CLog::Log(LOGERROR, "CDVDVideoCodecAndroidMediaCodec::GetOutputPicture(INFO_OUTPUT_FORMAT_CHANGED) ExceptionCheck: getOutputBuffers");
```
Comment (peak3d) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp]: ExceptionCheck

**Review by peak3d — COMMENTED** [review]
Looks good for me except the 2 minors.

Unfortunately this change bites completely with all the changes made in kodi-agile, what will be merged in the next time step by step into kodi master.

So question is how to proceed, any suggestion?

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp (hunk 1 above)
Comment (koying) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp]: m_codec is shared nowhere. Why do you see a need for a shared_ptr?

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp (hunk 2 above)
Comment (koying) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp]: No exceptions with the NDK C interface.

**Comment by koying** [comment]
@peak3d 
Even if the gains are marginal, that's totally way-to-go, and has been implemented that way in ffmpeg, too (it's so that I found out a mediacodec C interface existed ;) )

Re rebase, honestly, it's been months that we wait for agile, and > 1 week I forewarned about mediacodec changes. Do you expect me to wait for more weeks and take the rebase burden upon myself because you preferred to work on agile?

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp (hunk 1 above)
Comment (peak3d) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp]: You dived deeper into the code, only thing I see is that you have removed a shared_ptr (3 lines above) and use a normal pointer here. Are you sure that m_codec is not passed through DVDVideoPicture to the renderer?

B.t.w. no reason to get upset here, was just a question.

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp (hunk 2 above)
Comment (peak3d) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp]: In the log message is the word Exceptioncheck, wich does not make sense here.

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp (hunk 1 above)
Comment (koying) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp]: Well, m_codec is now an opaque pointer, not a class, so shared_ptr doesn't even make sense, here.
We could encapsulate in a class, I suppose, but I didn't have any crashes in the months I'm using this.

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp (hunk 2 above)
Comment (koying) [inline xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp]: Ah, indeed ;)

**Comment by MrMC** [comment]
Been waiting till our current work settled down to adapt MrMC to using the native interface for mediacodec. Indeed, this is the way forward.

**Comment by MartijnKaijser** [comment]
Jenkins build and merge

**Pull request merged**